File: 53. Maximum Subarray.cpp
class Solution {
public:
   int maxSubArray(vector<int>& nums) 
    {
        int localmax=nums[0];          
        int sum=nums[0];
        for(int i=1;i<nums.size();i++)
        {
            localmax=max(nums[i],localmax+nums[i]);
            sum=max(localmax,sum);
        }
        return sum;
    }
};

File: 53. MaximumSubarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int currentSum = 0;
        int maxSum = Integer.MIN_VALUE;

        for(int x: nums){
            currentSum += x;
            if(currentSum > maxSum) {
                maxSum = currentSum;
            }

            if(currentSum < 0){
                currentSum = 0; // restart subarray
            }
        }

        return maxSum;
    }
}
